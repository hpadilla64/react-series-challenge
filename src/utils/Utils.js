/**
 * A little helpers
 */

import ReactDOM from 'react-dom';

const Utils = {
    numberWithCommas: function (tgNum) {
        return tgNum.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    },
    debounce: function (func, wait, immediate) {
        let timeout;
        return function () {
            let context = this;
            let args = arguments;
            let later = function () {
                timeout = null;
                if (!immediate) func.apply(context, args);
            };
            let callNow = immediate && !timeout;
            clearTimeout(timeout);
            timeout = setTimeout(later, wait);
            if (callNow) func.apply(context, args);
        };
    },
    smoothScrollTo: function (element) {
        let headerHeight = document.getElementById('headerFixed').offsetHeight;
        let rect = ReactDOM.findDOMNode(element).getBoundingClientRect();
        let i = 10;
        let initialScroll = window.pageYOffset || document.documentElement.scrollTop;
        //console.log(initialScroll, ' | ', rect.top)
        let int = setInterval(function () {
            window.scrollTo(0, initialScroll + i);
            i += 20;
            if (i >= (rect.top - headerHeight)) {
                clearInterval(int)
            };
        }, 20);
    },
    jsonToURL: function (srcjson) {
        if (typeof srcjson !== "object") {
            if (typeof console !== "undefined") {
                console.log("\"srcjson\" is not a JSON object");
                return null;
            }
        }
        let u = encodeURIComponent;
        var urljson = "";
        var keys = Object.keys(srcjson);
        for (var i = 0; i < keys.length; i++) {
            if(typeof srcjson[keys[i]] === 'object' || typeof srcjson[keys[i]] === 'array') {
                srcjson[keys[i]] = JSON.stringify(srcjson[keys[i]]);
            }
            urljson += u(keys[i]) + "=" + u(srcjson[keys[i]]);
            if (i < (keys.length - 1)) {
                urljson += "&"
            };
        }
        return urljson;
    },
    slugfiy: function (str) {
        str = str.replace(/^\s+|\s+$/g, ''); // trim
        str = str.toLowerCase();

        // remove accents, swap ñ for n, etc
        let from = 'àáäâèéëêìíïîòóöôùúüûñç·/_,:;';
        let to = 'aaaaeeeeiiiioooouuuunc------';
        for (let i = 0, l = from.length; i < l; i++) {
            str = str.replace(new RegExp(from.charAt(i), 'g'), to.charAt(i));
        }

        str = str.replace(/[^a-z0-9 -]/g, '') // remove invalid chars
            .replace(/\s+/g, '-') // collapse whitespace and replace by -
            .replace(/-+/g, '-'); // collapse dashes

        return str;
    },
}

export default Utils;