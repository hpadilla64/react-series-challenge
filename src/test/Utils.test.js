import Utils from '../utils/Utils'

describe('test Utils', ()=> {
    it('Utils must have debounce', async () => {
        await Utils.debounce(()=>{
            expect(Utils).toHaveProperty('debounce');
        },1);
        
    })
});

describe('test debounce', ()=> {
    it('Debounce must be called', () => {
        const callback = jest.fn();
        expect(Utils.debounce(callback, 100)).toMatchSnapshot()
    })
});

describe('test debounce timer', () => {
    it('debounce', () => {
        const callback = jest.fn(() => ('test'));
        expect(Utils.debounce(callback, 100)).resolves.toBe('function');
    });
})

describe('test slugfy', ()=>{
    it('String to slug', ()=>{
        expect(Utils.slugfiy('Slug Test')).toBe('slug-test');
    })
});

describe('test json to url', ()=>{
    it('json in url query format', ()=>{
        expect(Utils.jsonToURL({test:'a',result: 'b'})).toBe('test=a&result=b');
    })
});

describe('test add comma to numbers', () => {
    it('test numberWithCommas', ()=>{
        expect(Utils.numberWithCommas(1000)).toBe('1,000');
    })
})

describe('test smoothScroll', () => {
    it('scroll to element', () => {
        const element = document.createElement('div');
        const header = document.createElement('header');
        header.id = 'headerFixed';
        header.style.heigh = '5em';
        document.body.appendChild(header);
        element.id = 'tg';
        document.body.appendChild(element);
        Utils.smoothScrollTo(element);
    })
})