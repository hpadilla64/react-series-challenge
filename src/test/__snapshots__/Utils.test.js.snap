// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test debounce Debounce must be called 1`] = `[Function]`;
